`storage`: add more log context to `continuous_batch_parser::consume_records`

WillemKauf opened the pull request:
The `ERROR` log line we see appearing in the `continuous_batch_parser` would be more helpful with added information about the batch consumer type and record batch header.

Add a new `ERROR` log line to output this additional context in the cold path.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

nvartolomei commented on src/v/storage/parser.cc:
```diff
@@ -203,7 +203,14 @@ void continuous_batch_parser::add_bytes_and_reset() {
-             get_stream(), sz, "parser::consume_records", _recovery)
+             get_stream(),
+             sz,
+             fmt::format(
```
This is a bit of a hot path. Unconditional Extra allocation doesn’t seem nice.

doesn’t this lead to a dangling string too? String is allocated but nothing holds it until reading future finishes.

WillemKauf replied:
Moved this to a new log line in the cold path

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--067c535b-7f8f-4f8d-b082-d48864fae62a-->
<!--TEST RESULTS-->
<!--30ce8ad1-27d2-4082-b877-619254137086-->
<!--TEST RESULTS-->
<!--4a8c57d1-e2dd-4e5f-ad58-ff5f72ccbdeb-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63192">build#63192</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cloud_storage_rpfixture.cloud_storage_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/63192#0195988e-9613-41a8-8403-65043aac9ef1 | FAIL | 0/2 |
| partition_balancer_simulator_test_rpunit.partition_balancer_simulator_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/63192#0195988e-9613-41a8-8403-65043aac9ef1 | FLAKY | 1/2 |
| rptest.tests.consumer_group_test.ConsumerGroupTest.test_group_lag_metrics | ducktape | https://buildkite.com/redpanda/redpanda/builds/63192#019598d8-f30f-4e6a-8cd1-cce545aac06a | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63192#019598e6-84fa-4a1f-83b4-d37ad2020dcc | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63192#019598d8-f30f-4e6a-8cd1-cce545aac06a | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63192#019598d8-f30e-4e20-92f2-f85c21a25b58 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63196">build#63196</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63196#01959a61-e39c-404f-ae52-9d0a90af1e60 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63196#01959a47-cc7b-4856-856a-2a4767f588d1 | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63196#01959a47-cc7a-4990-9acc-ab6c805b2d07 | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_upgrade_test.DatalakeUpgradeTest.test_upload_through_upgrade.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK | ducktape | https://buildkite.com/redpanda/redpanda/builds/63196#01959a61-e39d-4c78-ac0b-b6626b887978 | FAIL | 0/20 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63196#01959a61-e39b-411d-9726-dd7954415a9d | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63357">build#63357</a></summary>


| test_id                                                                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_higher_level_migration_api                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63357#0195af54-9ed4-4e54-a2f7-7220e16cf01f | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63357#0195af54-9ed4-4e54-a2f7-7220e16cf01f | FLAKY         | 1/2      |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_normalize.dataset_type=JSON                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63357#0195af54-9ed4-4e54-a2f7-7220e16cf01f | FLAKY         | 1/2      |
</details>
<!--/body-->

WillemKauf committed:
`storage`: add more log context to `continuous_batch_parser::consume_records`

This `ERROR` log line would be more helpful with added information about the
batch consumer type and record batch header.

Add a new `ERROR` log line to output these in the cold path.

WillemKauf committed:
`storage`: fix return type of `verify_read_iobuf`

This was being implicitly converted to `std::error_code`, and a future
commit will attempt to log the error using `to_string(parser_errc)`.

For that reason, the return type and a single call site where the log
line is to be added is amended with the proper type,
`result<iobuf, parser_errc>`, in order to properly log the error.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/63196-->
<!--54253b3e-32dd-434a-b375-098d5db76a04-->
<!--title-->
## Retry command for [Build#63196](https://buildkite.com/redpanda/redpanda/builds/63196)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/datalake_upgrade_test.py::DatalakeUpgradeTest.test_upload_through_upgrade@{"cloud_storage_type":1,"query_engine":"spark"}
```
<!--/body-->

dotnwat commented on src/v/storage/parser.cc:
```diff
@@ -204,15 +204,18 @@ ss::future<result<stop_parser>> continuous_batch_parser::consume_records() {
-          _consumer->consume_records(std::move(record.value()));
-          return _consumer->consume_batch_end().then([](stop_parser sp) {
-              return ss::make_ready_future<result<stop_parser>>(sp);
-          });
-      });
+      .then(
+        [this](result<iobuf, parser_errc> record)
```
> This was being implicitly converted to `std::error_code`.

why does only this call site need to change? seems like there are other callers of `verify_read_iobuf` that didn't change, and then it's not clear why the conversion was a problem in the first place.

WillemKauf replied:
> it's not clear why the conversion was a problem in the first place

I'm outputting the error type in the new log line added here, and `to_string(parser_errc)` [is defined here](https://github.com/redpanda-data/redpanda/blob/33dfcaf00394a9e884622aab675e630f10d6a5d4/src/v/storage/parser_errc.h#L27).

>why does only this call site need to change?

This is the only call site which has the return type written out explicitly (others use `auto`).

dotnwat replied:
> This is the only call site which has the return type written out explicitly (others use auto).

right, so in `consume_header` when `auto b = verify...` contains an error then we have
```
if (!b) {
  co_return b.error();
}
```
in a function that returns `result<stop_parser>` instead of `result<stop_parser, parser_errc>` but now it's getting a `parser_errc` instead `std::error_code` so presumably some new conversion is happening somewhere? to just say it was being implicitly converted doesn't seem complete enough to understand the _impact_ or _non impact_ of the change.

WillemKauf replied:
Good point, I am not correcting all of the sites of implicit conversion with this change- only at this level/call site where I want to log the error, other users still have the implicit conversion to `std::error_code`.

There's even [this awkward conversion](https://github.com/redpanda-data/redpanda/blob/33dfcaf00394a9e884622aab675e630f10d6a5d4/src/v/storage/parser.cc#L236) back to `parser_errc` in `parser.cc`...

I can go back through and correct all the call sites if you'd like to see that in this PR for completion's sake.

dotnwat replied:
it's not so much about "correcting" as it is about identifying the concern, and addressing it fully in the commit message and/or changing code. i only did the first part.

WillemKauf replied:
Amended the commit message to indict why only this call site was corrected.

dotnwat replied:
@WillemKauf i still don't feel like i understand the implications of this. the conversion escapes through public methods like consume_header() via b.error(). std::error_code and other error codes have different boolean polarities.

WillemKauf replied:
I see the concern. This was merged (accidentally) because I had auto-merge enabled and another reviewer approved.

If we don't fully grasp the implications of this change we can revert the commit, and add the log line without the error code being printed.

dotnwat replied:
we don't need to revert it, but we need explain how things are fine. because just saying that we were casting to std::error_code incorrectly feels like there is potential hidden semantic changes.

WillemKauf replied:
With this change, we have simply moved the location of the implicit cast of `parser_errc` to `std::error_code` from `verify_read_iobuf()` to the return statement in its three callers, namely 

* [`consume_header()`](https://github.com/WillemKauf/redpanda/blob/419f00436a4d49b2a2dfb5110f0a796b6ba225ae/src/v/storage/parser.cc#L101-L103)
* [`consume_records()`](https://github.com/WillemKauf/redpanda/blob/419f00436a4d49b2a2dfb5110f0a796b6ba225ae/src/v/storage/parser.cc#L211-L212) 
* [`run()`](https://github.com/WillemKauf/redpanda/blob/419f00436a4d49b2a2dfb5110f0a796b6ba225ae/src/v/storage/parser.cc#L308-L310)

All of these functions return a `result<>` type, which is default typedef'ed to an `std::error_code` for the `error_type` (template argument `S`).

https://github.com/redpanda-data/redpanda/blob/419f00436a4d49b2a2dfb5110f0a796b6ba225ae/src/v/base/outcome.h#L31-L35

The boolean polarity of the error code return type here doesn't matter- we are using `operator bool()` from `outcome::basic_result` to check for errors.

There are no outward functional changes to public users of these functions after the return type change/implicit cast location shift.

WillemKauf commented:
@dotnwat do you still feel there are any blockers for this PR?

nvartolomei reviewed (approved):
lgtm
